# 주석문

- **주석**은 코드의 실행과 무관한 문장을 기록하기 위한 기능이다

In [1]:
# 변수 선언
a = 10
print(a)

10


# Mark Down 실습

## 블록 인용문

- 블록 인용문 실습
> 첫 번째 블록 인용문
> > 두 번째 블록 인용문

## 순서 있는 목록
- 숫자와 마침표 사용

1. 첫 번째 목록
2. 두 번째 목록

## 줄바꿈
- \<br\> 태그 또는 공백 두 칸 다음에 엔터

문장1  
문장2

## 기울여 쓰기
- '*' ,  '_' 로 감싼 텍스트

- 문장 *기울여* 쓰기

## 취소선
- '~~'로 감싼 텍스트

- ~~취소 문장~~

## 글상자
- '`'로 감싼 텍스트

- `글상자` 테스트

## 링크
- 사이트 및 이메일 링크, \[제목](주소)

[네이버](https://www.naver.com)

## 내부 링크
- 내부 헤더로 이동하는 링크
- \[제목](#이동할 헤더 제목)
- 제목에 띄어쓰기가 있는 경우, '-'로 연결

[마크다운 실습](#Mark-Down-실습)

## 이미지 삽입
- \![제목](파일 이름 또는 웹 주소)

![파이썬 로고](python.jpg)

<img

## 코드 삽입
- \```코드이름``` 사이에 코드 블록 삽입
- 코드이름: python, html, javascript, java....

- 파이썬 코드 삽입
```python
import time
print('hello')
```

## 표
- 헤더 셀을 구분할 때 3개 이상의 --- 기호 사용
- 콜론(:)으로 정렬 방식을 지정

|Type-A|Predict|Type-B|
|:---:|:---:|:---:|
|Kind|Positive|Negative|
|Positive|10|20|
|Negative|10|20|

# 연산자와 데이터

In [9]:
# 산술 연산자

a = 10
b = 35

print(a+b)
print(a-b)
print(a*b)
print(a/b)
print(a//b)
print(b//a)

45
-25
350
0.2857142857142857
0
3


In [11]:
# + 연산자 오버로딩: 문자 결합. 단! 숫자와 문자열 결합은 불가
c = '사과'
d = 10

print(c+str(d))  # str(d): d변수의 데이터 타입을 문자열로 변환

사과10


In [12]:
# * 연산자 오버로딩: 문자열 반복
e = '하'
print(e*5)

하하하하하


In [13]:
# 나머지 연산자(mod): %
f = 25.5
g = 3

print(f % g)

1.5


In [14]:
# 거듭제곱 연산자: **
print(g ** 2)

9


In [16]:
# 논리 연산자

x = 1   #True
y = 0   #False
print(x and y)
print(x or y)

0
1


In [79]:
a = 10
b = 20
print(a,b)
print(a,b, sep=',')   # sep 파라미터: 하나의 프린트 함수에 여러 값을 출력할 때 사용할 구분 문자열
print(a, end=' ')    # end 파라미터: 값을 출력 후 끝에 출력할 문자열, 기본 값은 '\n' (new line)이다.
print(b)

10 20
10,20
10 20


[문제] A 회사의 주가가 56,000원이고 B 회사의 주가가 125,000원이라고 할 때, 어떤 사람이 A 주식 100주와 B 주식 35주를 가지고 있다면 그사람이 가지고 있는 주식의 총액을 계산하는 코드를 변수 선언 및 연산자를 사용하여 작성하시오.

In [29]:
A = 56000
B = 125000
total_money = (A * 100) + (B * 35)
# print('총 주식액: ', total_money)
print(f'총 주식액: {total_money:,}')

총 주식액: 9,975,000


[문제] 화시 온도(F)를 섭씨 온도(C)로 변환하는 다음 공식을 이용하여 화씨 온도가 50일 때, 섭씨 온도를 계산하시오.
- C = (F - 32) / 1.8

In [34]:
F = 50
C = (F - 32) / 1.8
print(f'화씨 온도가 {F}일 때 섭씨 온도는 {C}이다.')

화씨 온도가 50일 때 섭씨 온도는 10.0이다.


# 표준 입출력

In [35]:
a = int(input('밑변의 길이 입력: '))
b = int(input('높이의 길이 입력: '))
print('밑변의 길이가', a, '이고 높이가', b, '인 사각형의 넓이: ', a*b)

밑변의 길이 입력:  10
높이의 길이 입력:  20


밑변의 길이가 10 이고 높이가 20 인 사각형의 넓이:  200


# 제어문

In [36]:
score = float(input('점수 입력: '))
if score >= 90:
    print('합격')
elif 70 <= score < 90:
    print('재시험')
else:
    print('탈락')

점수 입력:  88


재시험


In [39]:
notEmptyString = 'test'

if notEmptyString:
    print('문자열이 비어 있지 않습니다.')
else:
    print('문자열이 비어 있습니다.')

문자열이 비어 있지 않습니다.


In [41]:
i = 20

if not i % 2:    #i % 2 == 0:
    print('짝수')
else:
    print('홀수')

짝수


[문제] 정수 하나를 입력 받아 5의 배수인지를 조사하여 그 결과를 출력하시오.

In [43]:
i = int(input('정수를 입력하세요: '))

if i % 5 == 0:
    print('5의 배수입니다.')
else:
    print('5의 배수가 아닙니다.')

정수를 입력하세요:  8


5의 배수가 아닙니다.


[문제] 화폐 단위 계산
사용자로부터 금액을 입력받아 화폐의 단위별 개수를 출력하는 프로그램을 작성하시오.
조건) 화폐의 종류: 만원, 천원, 백원, 십원
<실행 예>
금액을 입력하세요: 45630
만원: 4개
천원: 5개
백원: 6개
십원: 3개

In [54]:
money = int(input('금액을 입력하세요: '))
만원 = money // 10000
천원 = (money % 10000) // 1000 # (money - 만원*10000) // 1000
백원 = (money % 1000) // 100
십원 = (money % 100) // 10

print(f'만원: {만원}개')
print(f'천원: {천원}개')
print(f'백원: {백원}개')
print(f'십원: {십원}개')

금액을 입력하세요:  41520


만원: 4개
천원: 1개
백원: 5개
십원: 2개


In [55]:
num = int(input('금액을 입력하시오: '))

print('만원: ',  num // 10000 , '개')
print('천원: ', num%10000 // 1000, '개')
print('백원: ', num%1000 // 100, '개')
print('십원: ', num%100 // 10, '개')

금액을 입력하시오:  97415


만원:  9 개
천원:  7 개
백원:  4 개
십원:  1 개


## 삼항 연산자
- 결과값 변수 = 참인 경우 선택될 값 if 조건식 else 거짓인 경우 선택될 값

In [56]:
a = 10
b = 20
result = "a와 b는 같지 않다" if a != b else "a와 b는 같다"
print(result)

a와 b는 같지 않다


[문제] 대도시 판단 여부
[판단 조건] 
수도이면서 인구가 100만 이상인 도시
또는, 부자가 200만 이상인 도시

[실행 예시]
수도입니까? (수도: 1, 수도 아님: 0)  1
인구(단위:만) 150
부자의 수(단위:만) 50
대도시 여부: True

수도입니까? (수도: 1, 수도 아님: 0)  0
인구(단위:만) 90
부자의 수(단위:만) 210
대도시 여부: True

** 처리조건: 삼항 연산자 이용

In [67]:
isCapital = int(input('수도입니까? (수도:1, 수도아님:0 '))
citizens = int(input('인구(단위:만): '))
riches = int(input('부자의 수(단위:만): '))

result = True if (isCapital == 1, citizens >= 100) or riches >= 200 else False
print('대도시 여부', result)

수도입니까? (수도:1, 수도아님:0  0
인구(단위:만):  100
부자의 수(단위:만):  230


True


In [66]:
city = int(input('수도입니까?'))
population = int(input('총 인구(단위:만): '))
rich = int(input('부자의 수(단위:만): '))

rule1 = city == 1 and population >= 150
rule2 = rich >= 200

result = "True" if rule1 or rule2 else "False"
print(f'대도시 여부: {result}')

수도입니까? 1
총 인구(단위:만):  170
부자의 수(단위:만):  30


대도시 여부: True


[문제] 시간제 임금 계산
근무한 시간을 입력받아 다음과 같은 규칙에 맞게 임금을 계산하시오.
- 8시간까지는 시간당 9,900원이고, 8시간 초과분에 대해서는 1.5배를 지급한다.

[실행 예시]
근무시간을 입력하세요:  25
임금은 331,650원 입니다.

In [74]:
base_pay = 9900
hour = int(input('근무 시간을 입력하세요: '))
pay = 0  # 최종 임금을 저장할 변수

if hour > 8:
    pay = 8 * base_pay + (hour - 8) * base_pay * 1.5
else:
    pay = hour * base_pay
print(f'임금은 {pay}원입니다.')

근무 시간을 입력하세요:  25


임금은 331650.0원입니다.


In [78]:
time = int(input('근무 시간을 입력하세요: '))
over = (time - 8) * 9900 * 1.5
pay = 0
if time > 8:
    pay = (8 * 9900) + over
else:
    pay = time * 9900
print(f'임금은 {pay:,.0f}원입니다.')

근무 시간을 입력하세요:  25


임금은 331,650원입니다.


# 반복문

In [3]:
c = 1
while c <= 5:
    print(c, end=' ')
    c += 1
print('end')

1 2 3 4 5 end


In [4]:
sum = 0
for i in range(1,11):
    sum += i  # sum = sum + i
print(sum)

55


In [5]:
while True:  # 무한 반복
    s = input('영어 단어를 입력하세요. (종료:q)')
    if s == 'q': break    # if로 실행되는 문장이 1줄이면 옆에 이어서 쓰기 可
    print(s.upper())
print('end')

영어 단어를 입력하세요. (종료:q hello


HELLO


영어 단어를 입력하세요. (종료:q q


end


[문제] 1-100 사이의 값 중 짝수에 해당하는 값의 총 합을 구하는 프로그램을 작성하시오

In [21]:
sum = 0
for i in range(1, 101):
    if i % 2 != 0: continue
    sum += i
print(sum)

2550


In [24]:
# 파이썬에서 문자열은 집합형 데이터 [리스트]로 관리한다.
for s in 'abcdef':
    print(s, end=',')
print('\b')   #백스페이스

a,b,c,d,e,f


In [25]:
a = [1,2,3,4]  # a는 list 객체
result = []
for i in a:
    result.append(i*3)  #append()는 리스트 객체에 값을 추가하는 함수
print(result)

[3, 6, 9, 12]


In [26]:
# List comprehensiom
result = [i*3 for i in a ]
print(result)

[3, 6, 9, 12]


In [27]:
# List comprehension with if
result = [i*3 for i in a if not i%2]
print(result)

[6, 12]


In [30]:
# a 원소 중 짝수는 3배수하고 홀수는 원래의 값을 갖는 리스트를 생성
result = [i * 3 if not i%2 else i for i in a]      # 삼항 연산자 쓸 경우, if는 for의 왼쪽에 위치
print(result)

[1, 6, 3, 12]


In [ ]:
for i in range(5):
    for j in range(5):
        print(f'i={i}, j={j}')

## for - else 문
- while - else / for - else: 반복문이 break 등으로 중간에 빠져나오지 않고 끝까지 실행됐을 경우, else 문이 실행된다.

In [36]:
word_list = ['java', 'python', 'c++', 'pascal']
keyword = input('검색 단어를 입력하세요. ')
for word in word_list:
    if word == keyword:
        print('검색 결과:', word)
        break
else: print('검색 결과가 없습니다.')

검색 단어를 입력하세요.  word


검색 결과가 없습니다.


[문제] for문과 문자열 반복을 이용하여 다음 문자열을 출력하세요.
---------+---------+---------+---------+---------+

In [59]:
for i in range(5):
    print('-' * 9, end = '+')

---------+---------+---------+---------+---------+

In [60]:
for i in range(5):
    print('-' * 9, end = '')
    print('+', end='')

---------+---------+---------+---------+---------+

In [62]:
for i in range(5):
    for j in range(9):
        print('-', end='')
    print('+', end='')

---------+---------+---------+---------+---------+

[문제] 사용자로부터 입력받은 단의 구구단을 출력하는 프로그램을 작성하세요.
[실행예시]
단을 입력하세요: 3
3 x 1 = 3
3 x 2 = 6
...
3 x 9 = 27

In [ ]:
number = int(input('단을 입력하세요: '))
for i in range(1, 10, 1):
    result = number * i
    print(f'{number} x {i} = {result}')

In [ ]:
# 답 1
dan = int(input('단을 입력하세요: '))
for i in range(1, 10):
    print(dan, 'x', i, '=', dan * i)

In [83]:
# 답 2
dan = int(input('단을 입력하세요: '))
for i in range(1, 10):
    print(f'{dan} x {i} = {dan * i:<2d}')       # 자릿수 :2d, 02d (빈자리 0 채움), < 왼쪽 정렬

단을 입력하세요:  3


3 x 1 = 3 
3 x 2 = 6 
3 x 3 = 9 
3 x 4 = 12
3 x 5 = 15
3 x 6 = 18
3 x 7 = 21
3 x 8 = 24
3 x 9 = 27


[문제] 합계 구하기
정수를 입력받아 1부터 입력받은 정수까지의 합을 구하여 출력하는 프로그램을 작성하세요.
단, 반복문은 while을 이용할 것

In [92]:
num = int(input('정수를 입력하세요: '))
sum = 0   # 전체 합계를 저장
cnt = 0  # 반복 횟수를 저장
while True:
    cnt += 1
    if cnt > num: break
    sum += cnt
print(f'1부터 {num}까지의 합: {sum}')

정수를 입력하세요:  3


1부터 3까지의 합: 6


In [93]:
num = int(input('정수를 입력하세요: '))
sum = 0   # 전체 합계를 저장
cnt = 0  # 반복 횟수를 저장
while cnt < num:
    cnt += 1
    sum += cnt
print(f'1부터 {num}까지의 합: {sum}')

정수를 입력하세요:  3


1부터 3까지의 합: 6


[문제] 년도를 입력받아 윤년인지 평년인지를 판단하는 프로그램을 작성하세요.
프로그램은 반복적으로 수행되며 0을 입력 받으면 프로그램을 종료한다.
윤년 조건: 4로 나누어 떨어지고 100으로 나누어 떨어지지 않는 해, 또는 400으로 나누어 떨어지는 해
[실행 예시]
연도를 입력하세요(종료: 0)  2023
2023년은 평년입니다.

연도를 입력하세요(종료: 0) 2024
2024년은 윤년입니다.

In [129]:
while True:
    year = int(input('연도를 입력하세요(종료: 0)'))
    if year == 0: break
    result = '윤년' if (year % 4 == 0 and year % 100 != 0) or (year % 400 == 0) else '평년'
    print(f'{year}년은 {result}입니다.')
print('end')

연도를 입력하세요(종료: 0) 2024


2024년은 윤년입니다.


연도를 입력하세요(종료: 0) 2023


2023년은 평년입니다.


연도를 입력하세요(종료: 0) 0


end


In [127]:
while True:
    year = int(input('연도를 입력하세요(종료: 0)'))
    if year == 0: break
    if (year % 4 == 0 and year % 100 != 0) or (year % 400 == 0): 
        print(f'{year}년은 윤년입니다.')
    else: 
        print(f'{year}년은 평년입니다.')

연도를 입력하세요(종료: 0) 2024


2024년은 윤년입니다.


연도를 입력하세요(종료: 0) 2025


2025년은 평년입니다.


연도를 입력하세요(종료: 0) 0


[문제]  BMI 측정 프로그램
키와 몸무게를 입력받아 BMI 측정 결과를 출력하는 프로그램을 작성하세요.

[BMI 계산식]
BMI = 몸무게(kg) / 키(m)의 제곱
** BMI 비만도 판단 결과 지표
18.5 미만: 저체중
18.5 ~ 22.9: 정상
23 ~ 29.9: 비만 1단계
29.9 초과: 비만 2단계

[실행 예시]
키를 입력하세요(cm): 170
몸무게를 입력하세요(kg): 85
당신의 BMI는 비만 1단계 입니다.

In [144]:
height = int(input('키를 입력하세요(cm): '))
weight = int(input('몸무게를 입력하세요(kg)'))
BMI = weight / ((height/100) ** 2)
result = ''
if BMI < 18.5:
    result = '저체중'
elif 18.5 <= BMI <= 22.9:
    result = '정상'
elif 23 <= BMI <= 29.9:
    result = '비만 1단계'
else: result = '비만 2단계'
print(f'당신의 BMI는 {result} 입니다.')

키를 입력하세요(cm):  170
몸무게를 입력하세요(kg) 85


당신의 BMI는 비만 1단계 입니다.


In [143]:
height = int(input('키를 입력하세요(cm): '))
weight = int(input('몸무게를 입력하세요(kg)'))
BMI = weight / ((height/100) ** 2)

if BMI < 18.5:
    print('당신의 BMI는 저체중 입니다.')
elif 18.5 <= BMI <= 22.9:
    print('당신의 BMI는 정상 입니다.')
elif 23 <= BMI <= 29.9:
    print('당신의 BMI는 비만 1단계 입니다.')
else:
    print('당신의 BMI는 비만 2단계 입니다.')

키를 입력하세요(cm):  170
몸무게를 입력하세요(kg) 85


당신의 BMI는 비만 1단계 입니다.


# 문자열 처리

In [5]:
mystr = '학교종이 땡땡땡'
print(mystr[0])
print(mystr[-1])
# mystr[1] = '원'
for s in mystr:
    print(s)        # 한 글자씩 뽑아 냄

학
땡
학
교
종
이
 
땡
땡
땡


In [6]:
mystr = '''동해물과 백두산이 마르고 닳도록
하느님이 보우하사 우리나라 만세
무궁화 삼천리 화려 강산
대한 사람 대한으로 길이 보전하세'''
print(mystr)

동해물과 백두산이 마르고 닳도록
하느님이 보우하사 우리나라 만세
무궁화 삼천리 화려 강산
대한 사람 대한으로 길이 보전하세


In [9]:
hello = "나는 '안녕하세요'라고 인사합니다"
print(hello)
hello2 = "나는 \'안녕하세요\'라고 인사합니다"
print(hello2)
print('철수\t영희')

나는 '안녕하세요'라고 인사합니다
나는 '안녕하세요'라고 인사합니다
철수	영희


[문제] 파일 이름 슬라이싱
- 파일 이름: "20240315-163750.jpg"
[출력 결과]
촬영 날짜: 2024년 03월 15일
촬영 시간: 16시 37분
확장자: jpg

In [41]:
file = "20240315-163750.jpg"
print(f'촬영 날짜: {file[:4]}년 {file[4:6]}월 {file[6:8]}일')
print(f'촬영 시간: {file[9:11]}시 {file[11:13]}분')
print(f'확장자: {file[-3:len(file)]}')      # = {file[-3:]}
print(file[file.find('.'):])

촬영 날짜: 2024년 03월 15일
촬영 시간: 16시 37분
확장자: jpg
.jpg


In [42]:
# 문자열 뒤집기
str = 'hello'
print(str[::-1])

olleh


## 문자열 함수

In [44]:
s = '''생각이란 생각할수록 생각나므로
생각하지 말아야 할 생각은 생각하지 않으려고 하는 생각이
옳은 생각이라고 생각합니다.'''
print(s.count('생각'))

9


In [45]:
s = 'Python Programming'
print('a' in s) 
print('z' in s)

True
False


In [50]:
s = '짜장 짬뽕 탕수육'
print(s.split())                  # 기본적으로 공백 기준

s2 = '서울->대전->대구->부산'
print(s2.split('->'))

['짜장', '짬뽕', '탕수육']
['서울', '대전', '대구', '부산']


In [51]:
city_list = ['서울', '대전', '대구', '부산']
city_string = ','.join(city_list)
print(city_string)

서울,대전,대구,부산


In [54]:
s = '대한민국'
print('.'.join(s))      # 문자열 전체를 묶을 경우, 각 글자 = 원소

대.한.민.국


[문제] 주민등록번호 분석하기
- 임의의 주민등록번호를 입력받아 다음과 같이 생년과 성별을 추출하는 프로그램을 작성하세요.
- 성별 값은 1~4 사이의 값만 유효   성별오류
[실행 예시]
주민등록번호: 881231-1234567
88년생 남자

In [75]:
human = input('주민등록번호: ')
year = human[:2]
gender = human[7]
s_gender = ''
if gender in ['1','3']:
    s_gender = '남자'
elif gender in ['2','4']:
    s_gender = '여자'
else: gender = '성별오류'
print(f'{year}년생 {s_gender}')

주민등록번호:  001227-0000000


00년생 


In [76]:
human = input('주민등록번호: ')
if human[7] in ['1' , '3']:
    result = '남자'
elif human[7] in ['2','4']:
    result = '여자'
else: result = '성별오류'
print(f'{human[:2]} 년생  {result}')

주민등록번호:  001227-3512647


00 년생  남자


## 문자열 대입

In [10]:
str = "I eat %d apples"
print("I eat %d apples" %3)
print("I eat %s apples"%"five")
print("I eat %d%% apples" %95)

I eat 3 apples
I eat five apples
I eat 95% apples


In [11]:
month = 8
day = 15
anni = '광복절'
print('%d월 %d일은 %s이다' %(month, day, anni))

8월 15일은 광복절이다


In [18]:
number = 4
day = '삼'
print('나는 사과{}개를 먹고 {}일 동안 아팠다'.format(number, day))                #순서대로
print('나는 사과{0}개를 먹고 {1}일 동안 아팠다'.format(number, day))              #위치값
print('나는 사과{number}개를 먹고 {day}일 동안 아팠다'.format(number=7, day=3))   #key 값

나는 사과4개를 먹고 삼일 동안 아팠다
나는 사과4개를 먹고 삼일 동안 아팠다
나는 사과7개를 먹고 3일 동안 아팠다


In [22]:
print('내 이름은 {}입니다.'.format('홍길동'))
print('내 이름은 {:5s}입니다.'.format('홍길동'))    #{} 안에 삽입해 정렬 및 띄움
print('내 이름은 {:->5s}입니다.'.format('홍길동'))  #기호 삽입 시 띄우는 부분 기호로 채움

내 이름은 홍길동입니다.
내 이름은 홍길동  입니다.
내 이름은 --홍길동입니다.


[문제] 단어 수 세기
- 문자열 내 전체 단어가 몇 개인가?
- 문자열 내 'like' 단어의 개수가 몇 개인가?
- 문자열 내 가장 많이 포함되어 있는 단어는 무엇인가?
>- 1. 문자열 끝에 개행(\n) 문자를 공백 문자로 치환한다.   #replace 함수
>  2. 문자열을 단어로 나눈다.   #split 함수
>  3. 모든 단어를 소문자로 수정한다. #lower 함수

In [31]:
song = """Hello, it's me 
I was wondering if after all these years you'd like to meet
To go over everything
They say that time's supposed to heal ya
But I ain't done much healing
Hello, can you hear me?
I'm in California dreaming about who we used to be
When we were younger and free
I've forgotten how it felt before the world fell at our feet
There's such a difference between us
And a million miles
Hello from the other side
I must've called a thousand times
To tell you I'm sorry for everything that I've done
But when I call, you never seem to be home
Hello from the outside
At least I can say that I've tried
To tell you I'm sorry for breaking your heart
But it don't matter, it clearly doesn't tear you apart anymore
Hello, how are you?
It's so typical of me to talk about myself, I'm sorry
I hope that you're well
Did you ever make it out of that town where nothing ever happened?
It's no secret that the both of us are running out of time
So hello from the other side (Other side)
I must've called a thousand times (Thousand times)
To tell you I'm sorry for everything that I've done
But when I call, you never seem to be home
Hello from the outside (Outside)
At least I can say that I've tried (I've tried)
To tell you I'm sorry for breaking your heart
But it don't matter, it clearly doesn't tear you apart anymore
Ooh-ooh, anymore
Ooh-ooh-ooh, anymore
Ooh-ooh, anymore
Anymore
Hello from the other side (Other side)
I must've called a thousand times (Thousand times)
To tell you I'm sorry for everything that I've done
But when I call, you never seem to be home
Hello from the outside (Outside)
At least I can say that I've tried (I've tried)
To tell you I'm sorry for breaking your heart
But it don't matter, it clearly doesn't tear you apart anymore"""

song = song.lower().replace('\n',' ')
punctuation = ".,'`?!-()"
for p in punctuation:
    song = song.replace(p, '')
print('like 단어의 개수:' , song.count('like'))
word_list = song.split()             #매개 변수 넣지 않으면 기본적으로 공백 기준
print('전체 단어의 개수:' , len(word_list))

max_word_count = 0
max_word = ''
for word in word_list:
    word_count = word_list.count(word)
    if word_count > max_word_count:
        max_word_count = word_count
        max_word = word
print('가장 많이 나온 단어는 %s이며, 횟수는 %d이다.' %(max_word, max_word_count))

like 단어의 개수: 1
전체 단어의 개수: 333
가장 많이 나온 단어는 you이며, 횟수는 15이다.


<pre></pre>
    [문제] 영화 예매
    10개 좌석만 있는 극장의 예매 프로그램 만들기
    비어 있는 좌석은 0으로 표기하고, 예매가 된 좌석은 1로 표기한다.
    만약, 예약되어 있는 좌석을 예약하려고 했을 때 적절한 오류 메시지를 출력하고 다시 예약을 받는다.
    
    [실행 예시] 
    좌석을 예약하시겠습니까? (예약: 1, 종료: 0)
    현재의 예약 상태는 다음과 같습니다.
    --------------------------------------
     1 2 3 4 5 6 7 8 9 10
     -------------------------------------
     0 0 0 0 0 0 0 0 0 0
     
    몇 번째 좌석을 예약하시겠습니까? 2
    예약되었습니다.
    
    좌석을 예약하시겠습니까? (예약: 1, 종료: 0)
    --------------------------------------
     1 2 3 4 5 6 7 8 9 10
     -------------------------------------
     0 1 0 0 0 0 0 0 0 0
 <pre></pre>

In [73]:
# 좌석 정보를 저장하는 리스트abs
# 2번 좌석 예약 : seats[1] = 1
seats = [0,0,0,0,0,0,0,0,0,0]
while True:
    isContinue = int(input('좌석을 예약하시겠습니까? (예약: 1, 종료: 0)'))
    if isContinue == 0:
        break
    elif isContinue == 1:
        print('현재의 예약 상태는 다음과 같습니다.')
        print('------------------------')
        print(' 1 2 3 4 5 6 7 8 9 10 ')
        print('------------------------')
        for s in seats:
            print('%2d'%s, end='')
        while True:
            seat_num = int(input('몇 번째 좌석을 예약하시겠습니까?'))
            if seats[seat_num - 1] == 1:
                print('해당 좌석은 이미 예약된 좌석입니다.')
            else:
                seats[seat_num - 1] = 1
                print('예약되었습니다.')
                break

좌석을 예약하시겠습니까? (예약: 1, 종료: 0) 0


# 리스트 (List)

In [76]:
# 리스트 선언
a = []
b = list()                # 다른 데이터 타입을 리스트로 변환할 경우 함수 사용
c = [1,2,3]

In [78]:
# 슬라이싱
a = [1,2,3,4,5,6,7]
b = a[:2]
print(b)
c = a[:]
print(c)

[1, 2]
[1, 2, 3, 4, 5, 6, 7]


In [83]:
# 리스트 복사
a = [1,2,3,4,5,6,7]

# 얕은 복사
b = a
print(b)
b[2] = 20
print(a)

#깊은 복사
c = a[:]
c[0] = 200
print(a)

print(a is b)
print(a is c)

[1, 2, 3, 4, 5, 6, 7]
[1, 2, 20, 4, 5, 6, 7]
[1, 2, 20, 4, 5, 6, 7]
True
False


In [85]:
# 리스트의 연결(+), 반복(*)
a = [1,2,3]
b = [4,5,6]
print(a + b)
print(a * 3)

[1, 2, 3, 4, 5, 6]
[1, 2, 3, 1, 2, 3, 1, 2, 3]


In [90]:
# 리스트에 들어 있는 요소를 삭제하거나 수정하는 방법
a = [1,2,3,4,5]
a[3] = 40
print(a)

a[1:3] = []
print(a)

del a[1]
print(a)

[1, 2, 3, 40, 5]
[1, 40, 5]
[1, 5]


## 리스트의 함수들

In [93]:
a = [1, 5, 3, 6, 7]
a.sort()                    # sort 함수는 어떤 변수에 대입 X , 정렬의 결과 반환 X, 그저 정렬만 함
print(a)
a.sort(reverse=True)        # 내림차순
print(a)

[1, 3, 5, 6, 7]
[7, 6, 5, 3, 1]


In [96]:
a = [1, 3, 5, 2, 8, 9, 15, 29]
a.remove(5)                 # index로 삭제가 아닌 실제 값을 삭제하는 것
print(a)
temp = a.pop()              # 가장 마지막 데이터를 꺼낸 후 삭제
print(temp)
print(a)

[1, 3, 2, 8, 9, 15, 29]
29
[1, 3, 2, 8, 9, 15]


In [100]:
a = [1, 3, 4, 2, 1, 18, 22]
# 다음 함수는 리스트 객체의 함수가 아닌 파이썬의 내장 함수
print(len(a))
print(max(a))
print(min(a))
print(sum(a))
print(sum(a)/len(a))       # 평균값

7
22
1
51
7.285714285714286


In [102]:
ans = input('결제를 하시겠습니까?')
if ans in ['yes', 'y', 'ok', '네', '예']:
    print('구매해 주셔서 감사합니다.')
else:
    print('안녕히 가세요!')

결제를 하시겠습니까? y


구매해 주셔서 감사합니다.


## 이중 리스트
- 리스트의 요소로 리스트를 갖는 형태

In [107]:
l = [[1,2,3],
     [4,5,6],
     [7,8,9]]
print(l[0])
print(l[0][0])

for x in l:
    for y in x:
        print(y, end = ' ')
    print()                     # 줄바꿈

[1, 2, 3]
1
1 2 3 
4 5 6 
7 8 9 


## [:]과 deepcopy()의 차이

In [110]:
import copy   # 원하는 패키지를 사용하기 위함

matrix = [[1,2,3],
         [4,5,6],
         [7,8,9]]
# [:] 이용한 깊은 복사는 서브 리스트까지 깊은 복사를 수행하지 않는다.
# copy_m = matrix[:]
# 반면 deepcopy() 함수는 모든 요소에 대한 깊은 복사를 수행한다.
copy_m = copy.deepcopy(matrix)
print(id(matrix[0]))
print(id(copy_m[0]))

2246603231872
2246599756736


[문제] 성적 총점과 평균 구하기
학생 3명의 성적에 대해 각각 총점과 평균을 구하고, 전체 평균을 구하세요.
[실행 결과]
1번 학생: 총점: , 평균:
2번 학생: 총점: , 평균:
3번 학생: 총점: , 평균:
전체 평균:

In [131]:
score = [[88, 76, 92, 98],
         [65, 70, 58, 82],
         [82, 80, 78, 88]]
index = 1
total = 0  # 전체 모든 학생의 총점을 저장
all_subject = 0  # 전체 과목 수
for student in score:
    print(f'{index}번 학생: 총점: {sum(student)}, 평균: {sum(student) / len(student):.2f}')
    index += 1
    total += sum(student)
    all_subject += len(student)
print(f'전체 평균: {total/all_subject}')

1번 학생: 총점: 354, 평균: 88.50
2번 학생: 총점: 275, 평균: 68.75
3번 학생: 총점: 328, 평균: 82.00
전체 평균: 79.75


In [1]:
score = [[88, 76, 92, 98],
         [65, 70, 58, 82],
         [82, 80, 78, 88]]

stu = 0
stu_all = 0
for stu in score:
    stu_sum = sum(stu)
    stu_avg = sum(stu) / len(stu)
    stu_all += stu_avg
    print(f'{score.index(stu) + 1}번 학생: 총점: {stu_sum}, 평균: {stu_avg}')
print('전체 평균: ', stu_all / len(score))

1번 학생: 총점: 354, 평균: 88.5
2번 학생: 총점: 275, 평균: 68.75
3번 학생: 총점: 328, 평균: 82.0
전체 평균:  79.75


[문제] 일수 계산
1-12월 중 입력 받은 월이 몇 일까지 있는지를 계산하는 프로그램을 작성하세요.

[실행예시]
년도를 입력하세요: 2024
월을 입력하세요: 2
2024년 2월은 29일까지 있습니다.

[참고]
윤년 조건: 4로 나눠지고, 100으로 나워지지 않는 해 또는 400으로 나눠지는 해

In [146]:
daysInMonth = [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]

year = int(input('년도를 입력하세요: '))
month = int(input('월을 입력하세요: '))
days = 0
if month == 2:
    days = 29 if (not year % 4 and year % 100) or (not year % 400) else 28
else:
    days = daysInMonth[month-1]
print(f'{year}년 {month}월은 {days}일까지 있습니다.')

년도를 입력하세요:  2023
월을 입력하세요:  2


2023년 2월은 28일까지 있습니다.


In [143]:
daysInMonth = [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]

year = int(input('년도를 입력하세요: '))
month = int(input('월을 입력하세요: '))
if month == 2:
    if (year % 4 == 0 and year % 100 != 0) or (year % 400 == 0):
        print(f'{year}년 {month}월은 29일까지 있습니다.')
    else: print(f'{year}년 {month}월은 {daysInMonth[month-1]}일까지 있습니다.')
else:
    print(f'{year}년 {month}월은 {daysInMonth[month-1]}일까지 있습니다.')

년도를 입력하세요:  2025
월을 입력하세요:  2


2025년 2월은 28일까지 있습니다.


# 딕셔너리 (Dictionary)

In [152]:
# 딕셔너리 생성
x = {'name':'홍길동', 'age':'32'}
print(x['name'])
x['address'] = '서울시 강남구'
print(x['address'])
x['address'] = '서울시 강서구'
print(x)
print(x.get('birth', '생일 미등록'))

홍길동
서울시 강남구
{'name': '홍길동', 'age': '32', 'address': '서울시 강서구'}
생일 미등록


In [157]:
# 딕셔너리 데이터 순회
y = {'name':'전우치', 'phone':'010-1234-5678', 'age':28}
print(x.keys())
print(x.items())

for key in y.keys():
    print(key, y[key])
print('-'*15)
for key, value in y.items():
    print(key, value)

dict_keys(['name', 'age', 'address'])
dict_items([('name', '홍길동'), ('age', '32'), ('address', '서울시 강서구')])
name 전우치
phone 010-1234-5678
age 28
---------------
name 전우치
phone 010-1234-5678
age 28


In [163]:
score = {'홍길동':89, '임꺽정':88, '손오공':80, '전우치':67}
print('합계: ', sum(score.values()))
print(sorted(score.items(), key = lambda x:x[1]))  # lambda 변수:표현식   => 이름 없는 함수 정의 방식

합계:  324
[('전우치', 67), ('손오공', 80), ('임꺽정', 88), ('홍길동', 89)]


[문제] 단어 빈도수 구하기
text 문장 안에 포함된 각 단어의 빈도수를 구하여 dictionary 형태로 결과를 출력하세요.
단어는 key가 되고, 빈도수는 value가 된다.

In [1]:
song = """Hello, it's me 
I was wondering if after all these years you'd like to meet
To go over everything
They say that time's supposed to heal ya
But I ain't done much healing
Hello, can you hear me?
I'm in California dreaming about who we used to be
When we were younger and free
I've forgotten how it felt before the world fell at our feet
There's such a difference between us
And a million miles
Hello from the other side
I must've called a thousand times
To tell you I'm sorry for everything that I've done
But when I call, you never seem to be home
Hello from the outside
At least I can say that I've tried
To tell you I'm sorry for breaking your heart
But it don't matter, it clearly doesn't tear you apart anymore
Hello, how are you?
It's so typical of me to talk about myself, I'm sorry
I hope that you're well
Did you ever make it out of that town where nothing ever happened?
It's no secret that the both of us are running out of time
So hello from the other side (Other side)
I must've called a thousand times (Thousand times)
To tell you I'm sorry for everything that I've done
But when I call, you never seem to be home
Hello from the outside (Outside)
At least I can say that I've tried (I've tried)
To tell you I'm sorry for breaking your heart
But it don't matter, it clearly doesn't tear you apart anymore
Ooh-ooh, anymore
Ooh-ooh-ooh, anymore
Ooh-ooh, anymore
Anymore
Hello from the other side (Other side)
I must've called a thousand times (Thousand times)
To tell you I'm sorry for everything that I've done
But when I call, you never seem to be home
Hello from the outside (Outside)
At least I can say that I've tried (I've tried)
To tell you I'm sorry for breaking your heart
But it don't matter, it clearly doesn't tear you apart anymore"""

song = song.lower().replace('\n',' ')
punctuation = ',!()`?-.'
for p in punctuation:
    song = song.replace(p, '')                  # replace는 반환만 하기 때문에 같은 변수인 song에 대입
word_list = song.split()                        # split() = 공백 기준
result = {}                                     # = dict() 비어 있는 dictionary 생성 
for word in word_list:
    if  result.get(word) == None:
        result[word] = 1
    else: 
        result[word]  += 1                      #  result[word] = result[word] + 1
print(sorted(result.items(), key=lambda x:x[1], reverse = True))

#2
# f or word in word_list:
#    result[word] = result.get(word, 0) + 1
# print(sorted(result.items(), key=lambda x:x[1], reverse = True)) 

[('you', 15), ('to', 14), ('i', 12), ('that', 10), ('hello', 9), ("i've", 9), ("i'm", 8), ('it', 8), ('the', 8), ('but', 7), ('sorry', 7), ('anymore', 7), ('from', 6), ('tell', 6), ('for', 6), ('a', 5), ('other', 5), ('side', 5), ('thousand', 5), ('times', 5), ('outside', 5), ('tried', 5), ('everything', 4), ('say', 4), ('done', 4), ('can', 4), ('be', 4), ('when', 4), ('at', 4), ('of', 4), ("it's", 3), ('me', 3), ("must've", 3), ('called', 3), ('call', 3), ('never', 3), ('seem', 3), ('home', 3), ('least', 3), ('breaking', 3), ('your', 3), ('heart', 3), ("don't", 3), ('matter', 3), ('clearly', 3), ("doesn't", 3), ('tear', 3), ('apart', 3), ('about', 2), ('we', 2), ('and', 2), ('how', 2), ('us', 2), ('are', 2), ('so', 2), ('ever', 2), ('out', 2), ('oohooh', 2), ('was', 1), ('wondering', 1), ('if', 1), ('after', 1), ('all', 1), ('these', 1), ('years', 1), ("you'd", 1), ('like', 1), ('meet', 1), ('go', 1), ('over', 1), ('they', 1), ("time's", 1), ('supposed', 1), ('heal', 1), ('ya', 1), ("

In [ ]:
song = """Hello, it's me 
I was wondering if after all these years you'd like to meet
To go over everything
They say that time's supposed to heal ya
But I ain't done much healing
Hello, can you hear me?
I'm in California dreaming about who we used to be
When we were younger and free
I've forgotten how it felt before the world fell at our feet
There's such a difference between us
And a million miles
Hello from the other side
I must've called a thousand times
To tell you I'm sorry for everything that I've done
But when I call, you never seem to be home
Hello from the outside
At least I can say that I've tried
To tell you I'm sorry for breaking your heart
But it don't matter, it clearly doesn't tear you apart anymore
Hello, how are you?
It's so typical of me to talk about myself, I'm sorry
I hope that you're well
Did you ever make it out of that town where nothing ever happened?
It's no secret that the both of us are running out of time
So hello from the other side (Other side)
I must've called a thousand times (Thousand times)
To tell you I'm sorry for everything that I've done
But when I call, you never seem to be home
Hello from the outside (Outside)
At least I can say that I've tried (I've tried)
To tell you I'm sorry for breaking your heart
But it don't matter, it clearly doesn't tear you apart anymore
Ooh-ooh, anymore
Ooh-ooh-ooh, anymore
Ooh-ooh, anymore
Anymore
Hello from the other side (Other side)
I must've called a thousand times (Thousand times)
To tell you I'm sorry for everything that I've done
But when I call, you never seem to be home
Hello from the outside (Outside)
At least I can say that I've tried (I've tried)
To tell you I'm sorry for breaking your heart
But it don't matter, it clearly doesn't tear you apart anymore"""

song = song.lower().replace('\n',' ')
punctuation = ".,'`?!-()"
for p in punctuation:
    song = song.replace(p, '')
word = song.split(' ')
all = dict()
for i in word:
    if all.get(i) == None:
        all[i] = 1
    else:
        all[i] += 1
    all.get(i,[word.count(i)])
print(sorted(all.items(), key = lambda x:x[1], reverse = True))

# 집합(Set)

In [31]:
print(set('Hello Python'))

{' ', 'h', 'l', 'H', 'e', 'y', 'n', 'o', 'P', 't'}


In [34]:
a = {1,2,3,4,5,6}
b = {5,6,7,8}
print('합집합: ', a.union(b))
print('교집합: ', a.intersection(b))
print('차집합: ', a.difference(b))

합집합:  {1, 2, 3, 4, 5, 6, 7, 8}
교집합:  {5, 6}
차집합:  {1, 2, 3, 4}


# Comprehension

In [41]:
tmp = [1,2, 4, 6, 7, 33, 63, 7, 5, 10, 43]
a = [i for i in tmp]
print(a)

b = [i*3 for i in tmp]
print(b)

c = [i for i in tmp if not i%2]
print(c)

# 4행 7열 형태의 이중 리스트
double_list = [[0,0,0,0,0,0,0],
               [0,0,0,0,0,0,0],
               [0,0,0,0,0,0,0],
               [0,0,0,0,0,0,0]]
# print(len(double_list))
double_list = [ [0 for col in range(7)] for row in range(4)]
print(double_list)

[1, 2, 4, 6, 7, 33, 63, 7, 5, 10, 43]
[3, 6, 12, 18, 21, 99, 189, 21, 15, 30, 129]
[2, 4, 6, 10]
[[0, 0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0, 0]]


In [42]:
# set comprehension
s = [i for i in range(1,11)]
print(s)

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]


In [ ]:
# dictionary comprehension
result = [word:word_list.count(word) for word in word_list]
print(result)

[문제] 1-100까지 1씩 증가하는 데이터를 생성해서 3의 배수이거나 5의 배수인 데이터만 리스트에 포함시키는 프로그램을 작성하세요.

In [49]:
a = [ i for i in range(1,101) if not i%3 or not i%5 ]
print(a)

[3, 5, 6, 9, 10, 12, 15, 18, 20, 21, 24, 25, 27, 30, 33, 35, 36, 39, 40, 42, 45, 48, 50, 51, 54, 55, 57, 60, 63, 65, 66, 69, 70, 72, 75, 78, 80, 81, 84, 85, 87, 90, 93, 95, 96, 99, 100]


In [47]:
a = [ i for i in range(1,101) if (i%3 == 0 or i%5 == 0) ]
print(a)

[3, 5, 6, 9, 10, 12, 15, 18, 20, 21, 24, 25, 27, 30, 33, 35, 36, 39, 40, 42, 45, 48, 50, 51, 54, 55, 57, 60, 63, 65, 66, 69, 70, 72, 75, 78, 80, 81, 84, 85, 87, 90, 93, 95, 96, 99, 100]


# zip()

In [51]:
data1 = [1,2,3,4,5]
data2 = [10,20,30,40,50]
data3 = [100,200,300,400,500]

result = [ x for x in zip(data1, data2)]
print(result)

result2 = [ sum(x) for x in zip(data1, data2, data3)]
print(result2)

[(1, 10), (2, 20), (3, 30), (4, 40), (5, 50)]
[111, 222, 333, 444, 555]


In [54]:
yoil = ['월', '화', '수', '목','금', '토', '일']
food = ['갈비탕','순대국','칼국수','삼겹살']
for y, f in zip(yoil,food):
    print(f'{y}요일 메뉴:{f}')

월요일 메뉴:갈비탕
화요일 메뉴:순대국
수요일 메뉴:칼국수
목요일 메뉴:삼겹살


In [55]:
d = dict(zip(yoil, food))
print(d)

{'월': '갈비탕', '화': '순대국', '수': '칼국수', '목': '삼겹살'}


# enumerate()

In [58]:
yoil = ['월', '화', '수', '목','금', '토', '일']
print(list(enumerate(yoil)))

for i, item in enumerate(yoil, 1):
    print(i, item)

[(0, '월'), (1, '화'), (2, '수'), (3, '목'), (4, '금'), (5, '토'), (6, '일')]
1 월
2 화
3 수
4 목
5 금
6 토
7 일


In [62]:
color = ['빨강','주황','노랑','초록','파랑']            # 명목형 (범주형) 데이터
maping = {v:i for i, v in enumerate(color)}
print(maping)

{'빨강': 0, '주황': 1, '노랑': 2, '초록': 3, '파랑': 4}


# 함수

In [2]:
def restaurant(food, drink, desert):
    return {'food':food, 'drink':drink, 'desert':desert}

# 위치 인자를 이용한 함수의 호출
first = restaurant('스테이크', '포도주', '치즈케익')
print(first)

# 키워드 인자를 이용한 함수의 호출 (정의된 매개변수 순서에 상관없이 호출 가능)
second = restaurant(drink = '막걸리', desert = '파전', food = '잔치국수')
print(second)

{'food': '스테이크', 'drink': '포도주', 'desert': '치즈케익'}
{'food': '잔치국수', 'drink': '막걸리', 'desert': '파전'}


In [4]:
def restaurant(food = '연어초밥', drink = '사케', desert = '아이스크림'):      # 초기값 가지도록
    return {'food':food, 'drink':drink, 'desert':desert}
print(restaurant())
print(restaurant('돈부리'))

{'food': '연어초밥', 'drink': '사케', 'desert': '아이스크림'}
{'food': '돈부리', 'drink': '사케', 'desert': '아이스크림'}


In [6]:
def calcstep(begin, end, step=1):
    hap = 0
    for num in range(begin, end+1, step):
        hap += num
    return hap
print('1~10까지의 합 =',  calcstep(1,10))

1~10까지의 합 = 55


In [10]:
price = 1000   # 전역변수

def sales():
    global price   # 전역변수 사용
    price = 5000   # 지역변수
    b_price = 500  # 지역변수

sales()   # 함수의 호출 (사용)
print('sales() 함수 호출 후 price 값: ', price)
# print(b_price)   # 지역변수는 함수 밖에서 호출 불가능

sales() 함수 호출 후 price 값:  5000


[문제] 다음과 같은 결과를 출력하는 함수를 정의하세요.
<pre>
    print_star(5)
    *
    **
    ***
    ****
    *****
</pre>

In [27]:
def print_star(a):                           # return 값을 갖는 함수 = 처리 결과 받는 변수 선언 필요
    for i in range(1,a+1):
        print('*'*i)
print_star(5)

*
**
***
****
*****


In [28]:
def print_star(a):
    result = []
    for i in range(1,a+1):
        result.append('*'*i)
    return '\n'.join(result)
r = print_star(5)
print(r)

*
**
***
****
*****


In [30]:
balace = 100

def deposit(amount):
    global balace 
    balace += amount
    return balace

def withdraw(amount):
    result = False  # 출금 결과 저장 (True: 성공, False: 실패)
    global balace
    if balace >= amount:
        balace -= amount
        result = True
    return result

# result = deposit(10)
# print(result)

success = withdraw(1000)
if success: print('출금 성공')
else: print('잔고 부족')

잔고 부족


[문제] 함수를 이용한 계산기
<pre>
    이항사칙연산식을 매개변수로 받아 연산의 결과를 출력하는 함수를 정의하시오.
    - 함수 형식: calculator(exp), 반환 값 없음
    - exp: 사용자로부터 입력받은 문자열 형태의 연산식
    - +, -, *, / 외 연산자 사용 시 오류 메시지 출력
    - 나누기 연산에서 부모의 값이 0인 경우, 오류 메시지 출력
    (실행 예시)
    계산식을 입력하세요: 10+20     => 문자열 형태  input()   - 함수 안에서 분해해서 사칙연산식 만들어 줘야됨..
    10+20=30.00

    계산식을 입력하세요: 10a20
    연산자 오류
</pre>

In [84]:
# 함수의 정의
def calculator(exp):
    operator = ''  # 발견된 연산자 문자열을 저장
    for op in ['+','-','*','/']:
        op_index = exp.find(op)
        if op_index != -1:  # 연산식 내에서 찾고자 하는 연산자가 발견이 됐을 경우
            operator = op
            break
    if op_index != -1:
        value1 = float(exp.split(operator)[0])
        value2 = float(exp.split(operator)[1])

        if operator == '+':
            print(f'{exp} = {value1 + value2:.2f}')
        elif operator == '-':
            print(f'{exp} = {value1 - value2:.2f}')
        elif operator == '*':
            print(f'{exp} = {value1 * value2:.2f}')
        elif operator == '/':
            if value2 != 0:
                print(f'{exp} = {value1 + value2:.2f}')
            else: print('0으로 나눌 수 없습니다.')
    else:
        print('연산자 오류!')

d = input('계산식을 입력하세요: ')
# 함수의 사용(호출)
calculator(d)

계산식을 입력하세요:  10+20


10+20 = 30.00


In [89]:
def calculator(exp):
    operator = ''   # 발견된 연산자 문자열을 저장
    for op in ['+','-','*','/'] :
        op_index = exp.find(op)
        if op_index != -1: # 연산식 내에서 찾고자 하는 연산자가 발견이 됐을 경우
            operator = op
            break
    if op_index == -1:
        print('연산자 오류')
        return     # 함수의 실행을 끝내겠다.
        
    value1 = float(exp.split(operator)[0])
    value2 = float(exp.split(operator)[1])

    result = 0.0
    
    if operator == '+':
        result = value1 + value2
    elif operator == '-':
        result = value1 - value2
    elif operator == '*':
        result = value1 * value2
    elif operator == '/':
        if value2 != 0:
            result = value1 / value2
        else:
            print('0으로 나눌 수 없습니다.')
            return

    print(f'{exp}={result:.2f}')

d = input('계산식을 입력하세요:')
calculator(d)            

계산식을 입력하세요: 10+20


10+20=30.00


In [98]:
# 함수의 정의abs
def calculator(exp):
    pro = '+-*/'
    for i in pro:
        if i in exp:
            num = d.split(i)
            value1 = float(num[0])
            value2 = float(num[1])
            result = ''
            if i == '+':
                result = value1 + value2
            elif i == '-':
                result = value1 - value2
            elif i == '*':
                result = value1 * value2
            else: 
                if value2 != 0:
                    result = value1 / value2
                else: print('0으로 나눌 수 없습니다.')
            print(f"{d}={result:.2f}")
            break
    else: print('연산자 오류!')
        
d = input('계산식을 입력하세요: ')
# 함수의 사용 (호출)
calculator(d)

계산식을 입력하세요:  1a1


연산자 오류!


## eval() - 동적 코드 실행
- eval 함수는 문자열 형태로 된 파이썬 표현식을 평가하여 그 결과를 리턴한다.

In [102]:
result = eval('1 + 2 * 3')
print(result)

a = 2
print(eval('a + 3'))

data = '["서울", "대전", "광주"]'
for c in eval(data):
    print(c)

7
5
서울
대전
광주


# *args와 **kwargs 매개변수

In [108]:
def test(*args):
    print('매개변수 값:',args)

test()
test(1)
test(1,2,3,'a','b','c')

매개변수 값: ()
매개변수 값: (1,)
매개변수 값: (1, 2, 3, 'a', 'b', 'c')


In [107]:
a = (1)
print(type(a))
b = (1,)                    # 튜플은 원소가 하나일 때 뒤에 반드시 ,로 찍어 줘야 함
print(type(b))

<class 'int'>
<class 'tuple'>


In [111]:
def test(a, b, *args):      # 위치 인수, 가변 매개변수 같이 쓸 경우 가변을 위치 뒤로
    print('a->', a)
    print('b->', b)
    print('args->',args)

test(10, 20)
test(10, 20, 30)

a-> 10
b-> 20
args-> ()
a-> 10
b-> 20
args-> (30,)


In [114]:
def test(a, **kwargs):
    print(a, kwargs)

test(10)
test(10, b=20)

10 {}
10 {'b': 20}


In [4]:
def calcstep(**kwargs):
    begin = kwargs['begin']
    end = kwargs['end']
    step = kwargs['step']

    tap = 0
    for n in range(begin, end+1, step):
        tap += n

    return tap

print('1-10 합계: ', calcstep)

1-10 합계:  <function calcstep at 0x00000251C81E7880>


In [ ]:
dic = {}
dic['a']
dic,get('a')

# Filert, Map, Reduce

In [115]:
def fluck(s):
    return s < 60                           # s가 60 미만이면 True, 아니면 False

score = [45, 89, 72, 53, 94]
for s in filter(fluck, score):
    print('낙제 점수:',s)

낙제 점수: 45
낙제 점수: 53


In [117]:
def half(s):
    return s / 2

score = [45, 89, 72, 53, 94]
for s in map(half, score):
    print(s, end=' ')

22.5 44.5 36.0 26.5 47.0 

In [120]:
from functools import reduce

# reduce 함수에 사용할 집계 함수는 두 개의 인수를 받는다.
# 첫 번째 매개변수는 누적자(accumulator)가 되고
# 두 번째 매개변수는 현재 값이 넘어오게 된다.
def hap(a, b):
    return a + b
    
score = [45, 89, 72, 53, 94]
print(reduce(hap, score))

alphabat = ['H', 'e', 'l', 'l', 'o']
print(reduce(hap, alphabat))

353
Hello


# Lambda

In [122]:
def myfunc(x, y):
    return x + y

result = myfunc(10, 20)
print(result)

30


In [123]:
result = lambda x, y : x+y
print(result(10,20))

30


In [126]:
data = {'a':10, 'b':2, 'c':13, 'd':7}
print(sorted(data))               # dic key의 사전적 순서로 정렬
print(sorted(data.items(), key = lambda x:x[1]))    # 값을 기준으로 정렬 (기본 오름차순)

['a', 'b', 'c', 'd']
[('b', 2), ('d', 7), ('a', 10), ('c', 13)]


In [129]:
def squared(s):
    return s**2 

a = [1, 2, 3, 4, 5]
print(list(map(lambda x:x**2, a)))
print(list(map(squared, a)))

[1, 4, 9, 16, 25]
[1, 4, 9, 16, 25]


In [130]:
x = [1, 3, 5, 7, 9]
y = [2, 4, 6, 8, 10]

print(list(map(lambda x1, y1:x1+y1, x, y)))

[3, 7, 11, 15, 19]


In [132]:
k = list(range(1, 11))

print(list(map(lambda x: x**2 if not x%2 else x, k)))

[1, 4, 3, 16, 5, 36, 7, 64, 9, 100]


In [142]:
score = [45, 30, 90, 100, 67]
print(list(filter(lambda x:x<60, score))) # 모든 객체는 데이터 형태 아주 다양
# print로 출력 시 그 형태에 맞게 자동 출력 불가능해, 함수가 정의가 되어 있지 않은 경우, 미리 기입해 줌

[45, 30]


In [135]:
def calc(op, a, b):
    op(a, b)

def add(a, b):
    print(a+b)

def multi(a, b):
    print(a*b)

calc(add, 1, 2)
calc(multi, 2, 3)

3
6


In [136]:
calc(lambda a,b:print(a/b), 5, 2)

2.5


In [140]:
help('a'.join)

Help on built-in function join:

join(iterable, /) method of builtins.str instance
    Concatenate any number of strings.
    
    The string whose method is called is inserted in between each given string.
    The result is returned as a new string.
    
    Example: '.'.join(['ab', 'pq', 'rs']) -> 'ab.pq.rs'



# 표준 모듈

In [2]:
import time

t = time.time()
print(t)                         # <- 1970.1.1.0.0부터 호출한 시간까지의 초
print(time.ctime())              # 사람이 읽기 쉽게 ctime

1711071025.737521
Fri Mar 22 10:30:25 2024


In [5]:
import time

start = time.time()
sum = 0
for i in range(1, 100000001):
    sum += i
end = time.time()
print('1-1억까지 합계 계산 총 소요시간:', (end-start))

1-1억까지 합계 계산 총 소요시간: 11.576533794403076


In [8]:
import time

now = time.localtime()         # 현재 컴퓨터 위치 지역 정보의 시간
# print(now)
print('%d년 %d월 %d일'%(now.tm_year, now.tm_mon, now.tm_mday))
print('%d시 %d분 %d초'%(now.tm_hour, now.tm_min, now.tm_sec))

2024년 3월 22일
10시 40분 59초


In [9]:
import time

print('이 문장은')
time.sleep(2)               # 매개변수 = 초 단위
print('2초 단위로')
time.sleep(2)
print('출력됩니다.')

이 문장은
2초 단위로
출력됩니다.


In [12]:
import random

random.seed(0)                             # 난수 발생 기존값 고정, 항상 같은 값 추출 
for i in range(5):
    print(random.randint(1,10))

7
7
1
5
9


In [13]:
import random 

food = ['짜장면','짬뽕','탕수육','군만두']
print(food)
random.shuffle(food)
print(food)

['짜장면', '짬뽕', '탕수육', '군만두']
['짜장면', '탕수육', '짬뽕', '군만두']


In [14]:
print(random.choice(food))

군만두


In [15]:
print(random.sample(food, 2))

['짬뽕', '군만두']


[문제] 숫자 맞추기 게임
- 1-100 사이의 난수를 맞추는 게임
- 입력 값과 난수 값을 비교하여 어떤 값이 더 큰지 힌트를 알려준다.
- 난수를 맞추면 시도 횟수 출력과 함께 프로그램을 종료한다.

In [20]:
import random
num = random.randint(1,100)
time = 0
while True:
    guess = int(input('입력: '))
    time = time + 1
    if guess == num: break
    else:
        if guess > num:
            print('숫자가 더 작습니다.')
        else: print('숫자가 더 큽니다.')
        continue
print(f'정답입니다. 시도 횟수: {time}')

입력:  50


숫자가 더 작습니다.


입력:  30


숫자가 더 작습니다.


입력:  20


숫자가 더 작습니다.


입력:  10


숫자가 더 큽니다.


입력:  15


숫자가 더 큽니다.


입력:  17


숫자가 더 큽니다.


입력:  19


숫자가 더 작습니다.


입력:  18


정답입니다. 시도 횟수: 8


[문제] 로또 번호 생성기
- 1-45 사이의 숫자 6개 뽑기
- 단, 숫자를 뽑을 때 중복된 숫자는 사용할 수 없음
- 결과는 오름차순으로 정렬해서 출력한다.

In [23]:
import random
lotto = set()   # 원소의 중복을 허용하지 않는다.
while True:
    lotto.add(random.randint(1,45))
    if len(lotto) == 6: break
print(sorted(list(lotto)))

[5, 7, 22, 31, 36, 44]


In [24]:
lotto = []
while True:
    num = random.randint(1,45)
    if num not in lotto:
        lotto.append(num)
    if len(lotto) == 6: break
lotto.sort()
print(lotto)

[7, 21, 23, 28, 40, 41]


In [26]:
num_list = list(range(1,46))
lotto = random.sample(num_list, 6)
print(sorted(lotto))

[1, 4, 6, 26, 36, 43]


# File I/O

In [27]:
# 파일 생성
f = open('myfile.txt', 'w')
f.close()
print('End')

End


In [6]:
# 파일 생성 후 문자열 쓰기
f = open('myfile.txt', 'w')
f.write('Hello World')
f.close()
print('End')

End


In [8]:
# 파일 생성 후 한글이 포함된 문자열 쓰기
f = open('myfile.txt', 'w', encoding = 'utf-8')
f.write('안녕하세요!')
f.close()
print('End')

End


In [ ]:
f = open('song .txt', 'r')
data = f.readlines()
print(data)
f.close()

In [ ]:
f = open('song .txt', 'r')
while True:
    line = f.readline()
    if not line: break
    print(line, end='')
f.close()

In [13]:
# 데이터 추가
f = open('song .txt', 'a')
f.write('======================')
f.close()
print('end')

end


## with문 사용
- with문을 사용하면 with 블럭을 벗어나는 순간 열린 파일 객체가 자동으로 닫힌다.

In [ ]:
with open('song .txt','r') as f:
    print(f.read())

In [15]:
poem = """Program is fun
When the word is done
If you wanna make your work also fun
use Python!
"""
with open('poem.txt', 'w') as f:
    f.write(poem)

In [16]:
f = open('poem.txt', 'r')
f.seek(3,0)
print(f.read())
f.close()

gram is fun
When the word is done
If you wanna make your work also fun
use Python!



In [18]:
original = open('python.jpg','rb')
target = open('copy_python.jpg','wb')
data = original.read()
target.write(data)
original.close()
target.close()
print('End')

End


[문제] 학생들의 성적 합계 평균 구하여 저장하기
- score.txt 파일에 저장되어 있는 학생들의 점수를 읽어서 합계와 평균을 구한 후, result.txt 파일에 결과를 저장하세요.
- (저장 형식)
  이름  합계  평균
  홍길동 270 89.0
  임꺽정 291 93.0

In [92]:
source = open('score.txt', 'r', encoding = 'utf-8')
target = open('result.txt', 'w', encoding = 'utf-8')
target.write('이름\t합계\t평균\t\n')
while True:
    score = source.readline()
    if not score: break
    data = score.split()
    total = int(data[1]) + int(data[2]) + int(data[3])
    avg = total / 3
    target.write('%3s\t%4d\t%.2f\n'%(data[0],total,avg))
source.close()
target.close()
print('end')

end


# File 과 Dictionary

## os 모듈

In [107]:
import os

path = r'G:\내 드라이브\DIMA4\01.Python'
os.chdir(path)
print(path)

G:\내 드라이브\DIMA4\01.Python


In [108]:
# 현재 작업 디렉토리 및 경로 변경

c_path = os.getcwd()
print('현재 작업 디렉토리: ',c_path)
os.chdir('..')
print('상위 디렉토리 이동')
c_path = os.getcwd()
print('현재 작업 디렉토리: ',c_path)

현재 작업 디렉토리:  G:\내 드라이브\DIMA4\01.Python
상위 디렉토리 이동
현재 작업 디렉토리:  G:\내 드라이브\DIMA4


In [110]:
# 파일 및 디렉토리 목록 확인
path = r'g:\내 드라이브\DIMA4\01.Python'
os.chdir(path)

list = os.listdir()
for item in list:
    if os.path.isfile(item):
        print('File: ',item,',',os.path.getsize(item),'bytes')
    else:
        print('Dir: ',item)

Dir:  .ipynb_checkpoints
File:  Python_basic.ipynb , 122304 bytes
File:  python.jpg , 19117 bytes
File:  Cording_test.ipynb , 32425 bytes
File:  practice.ipynb , 17148 bytes
File:  sysargv.py , 100 bytes
File:  myfile.txt , 16 bytes
File:  song .txt , 1818 bytes
File:  poem.txt , 90 bytes
File:  copy_python.jpg , 19117 bytes
File:  score.txt , 96 bytes
File:  result.txt , 133 bytes


## glob 모듈

In [113]:
import os
import glob

os.chdir('..')
print('현재 작업 디렉토리:',os.getcwd())
# 현재 디렉토리 내 모든 디렉토리 출력: **/, */
for f in glob.glob('**/'):
    print(f)

현재 작업 디렉토리: g:\내 드라이브\DIMA4
01.Python\


In [115]:
# 현재 디렉티로 내 디렉토리와 파일 출력: **, *
os.chdir('./01.Python/')
for f in glob.glob('**'):
    print(f)

Python_basic.ipynb
python.jpg
Cording_test.ipynb
practice.ipynb
sysargv.py
myfile.txt
song .txt
poem.txt
copy_python.jpg
score.txt
result.txt


In [116]:
os.chdir(r'G:\내 드라이브\DIMA4\01.Python')
for f in glob.glob('*.ipynb'):
    print(f)

Python_basic.ipynb
Cording_test.ipynb
practice.ipynb


# 예외

## 예외 발생 상황

In [120]:
# indexError
a = [1, 2, 3, 4, 5]
# print(a[5])

# ValueError
b = '1234b'
# c = int(b)

# ZeroDivisionError
# print(10/0)

# FileNotFoundError
# f = open('not_exists.txt', 'r')

## 예외 처리 (try ~ except)

In [122]:
try:
    data = 10 / 0
except:
    print('0으로 나눌 수 없습니다.')

print('end')

0으로 나눌 수 없습니다.
end


## 예외 처리 (try ~ except ~ except)

In [125]:
try:
    a = [1, 2, 3]
   # print(10/0)
    print(a[3])
except ZeroDivisionError as ze:
    print('0으로 나눌 수 없습니다.')
    print(ze)
except IndexError as ie:
    print('리스트의 범위를 벗어났습니다.')
    print(ie)
print('End')

리스트의 범위를 벗어났습니다.
list index out of range
End


In [126]:
try:
    a = [1, 2, 3]
    print(10/0)
    print(a[3])
except (ZeroDivisionError, IndentationError) as e:
    print(e)

division by zero


## 예외 처리 (try ~ except ~ finally)

In [128]:
try:
    a = [1, 2, 3]
    print(a[3])
except IndexError:
    print('리스트의 범위를 벗어났습니다.')
finally:
    print('항상 실행되는 코드')
print('End')

리스트의 범위를 벗어났습니다.
항상 실행되는 코드
End


In [129]:
def inputNum():
    num = int(input('양수 입력:'))
    if num < 0:
        raise ValueError('양의 정수만 입력')
    return num

try:
    num = inputNum()
    print(num)
except ValueError as e:
    print(e)

양수 입력: -3


양의 정수만 입력


In [130]:
a = 10
print(type(a))

b = 'abc'
print(type(b))

<class 'int'>
<class 'str'>


# Class

In [ ]:
 class FourCal:          # 클래스 이름 선언 시 일반적으로 대문자 (묵시적)
     pass 
    # 생성자: 클래스의 객체를 생성할 때 멤버변수를 초기화하기 위한 함수 (속성 = 멤버변수)
    # def __init__(self, first, second):
    #   self.first = first     # FourCal 클래스의 멤버 변수
    #    self.second = second   # FourCal 클래스의 멤버 변수

a = FourCal()
# a = FourCal(3, 4)
# print(a.first)
# print(a.second)

In [11]:
class Account:
    # Account 클래스의 속성(멤버변수)을 특정 값으로 초기화 하기 위한 생성자
    def __init__(self, accNo, balance, owner):
        self.accNo = accNo
        self.balance = balance
        self.owner = owner

    # Account 객체가 가지고 있는 잔고 값을 조회하는 함수
    def getBalance(self):               # 매개변수 갖지 않더라도 자기 자신을 갖는 참조변수는 선언 필수
        return self.balance

    def accountinfo(self):
        print(f'계좌번호:{self.accNo}, 소유주:{self.owner}')

# => 세 가지의 속성과 두 가지의 행동방식을 갖는 Account 생성   [붕어빵 틀]
# => 사용 위해 객체 생성 필요  (= a1, a2)

a1 = Account('11-11', 500, '홍길동')
a2 = Account('11-12', 600, '손오공')

a1.accountinfo()
print('잔고: ', a1.getBalance())
a2.accountinfo()
print('잔고: ', a2.getBalance())

계좌번호:11-11, 소유주:홍길동
잔고:  500
계좌번호:11-12, 소유주:손오공
잔고:  600


# 상속(Inheritance)

In [14]:
class Human:
    def __init__(self, name, age):
        self.name = name
        self.age = age
        
    def intro(self):
        print(f'저는 {self.age}살 {self.name}입니다.')

kim = Human('김철수', 29)
kim.intro()

lee = Human('이길동', 19)
lee.intro()

저는 29살 김철수입니다.
저는 19살 이길동입니다.


In [17]:
class Student(Human):
    def __init__(self, name, age, stdNo):
        super().__init__(name, age)        # 부모 클래스(Human)의 생성자 호출
        self.stdNo = stdNo

    def study(self):
        print('Study Hard-!!')

    # method overriding(매서드 재정의)
    # 상속 관계에서 부모 클래스에 정의된 매서드[함수]를 자식 클래스의 상황에 맞게 재정의 하는 것
    # => 부모에 정의된 것과 똑같이 정의해야 함, 다만 선언되어지는 실행 내용만 달라짐a
    
    def intro(self):
        super().intro()    # 부모 클래스(Human)의 intro() 매서드 호출
        print(f'학번은 {self.stdNo}입니다.')
        

kim = Student('김한별', 21, '2024')
kim.study()
kim.intro()

Study Hard-!!
저는 21살 김한별입니다.
학번은 2024입니다.


# 은닉 (Encapsulation)

- property를 이용한 setter / getter
    - 멤버 변수에 접근할 때 getter / setter 메서드를 이용하지 않고 직접 멤버 변수를 호출하여 사용 가능
    - 하지만 내부적으로는 getter / setter 메서드를 통해 값을 읽고 쓰게 된다.

In [24]:
class Date:
    def __init__(self, month):
        # 가급적 멤버 변수의 이름을 어렵게 설정 
        self.__inner_month = month

    def getMonth(self):                             # 읽기
        print('getMonth() call')
        return self.__inner_month

    def setMonth(self, month):                      # 값 변경, 쓰기
        if 1 <= month <= 12:
            self.__inner_month = month
        print('setMonth() call')

    month = property(getMonth, setMonth)            # 외부에 노출되는 것 = month 

today = Date(3)
# print(today.month)        # 값을 읽는 행위
# today.month = 12            # 값을 쓰는 행위
# print(today.month)
today.__inner_month = 10  # 은닉된 멤버 변수에 새로운 값을 쓴 것이 아니라 새로운 멤버 변수를 추가한 행위
# print(today.month)
print(today.__dict__)

{'_Date__inner_month': 3, '__inner_month': 10}


- @property 데코레이터를 이용한 setter / getter
    - getter 메서드 앞: @property
    - setter 메서드 앞: @<프로퍼티 이름>.setter

In [27]:
class Date:
    def __init__(self, month):
        self.__inner_month = month

    @property
    def month(self):
        print('getMonth() call')
        return self.__inner_month

    @month.setter
    def month(self, month):
        if 1 <= month <= 12:
            self.__inner_month = month
        print('setMonth() call')

today = Date(3)
print(today.month)
today.month = 4

getMonth() call
3
setMonth() call


# 사용자 정의 예외 클래스

In [28]:
class MyException(Exception):
    def __init__(self, message):
        super().__init__(message)

raise MyException("test")

MyException: test

# 다형성 (Polymorphism) - 오버라이딩

In [41]:
class MyDate:
    def __init__(self, year, month, day):
        self.year = year
        self.month = month
        self.day = day

    # object 클래스에 있는 __str__ 메서드 재정의
    def __str__(self):
        return f'{self.year}-{self.month}-{self.day}'

    def __eq__(self, other):
        return self.year == other.year and self.month == other.month and self.day == other.day

    def __gt__(self, other):
        if self.year > other.year:
            return True
        elif self.year == other.year and self.month > other.month:
            return True
        elif self.year == other.year and self.month == other.month and self.day > other.day:
            return True
        else: return False

d1 = MyDate(2024, 3, 26)
d2 = MyDate(2024, 3, 25)
# d2 = d1
# print(d)
# print(d1 == d2)
print(d1 > d2)

True


# Banking System

In [12]:
from bank import Account, BankManager, InvalidTransactionException

a1 = Account('홍길동', '11-11', 500)
print(a1.showAccountInfo())

a2 = Account('김철수', '11-12', 1000)
print(a2.showAccountInfo())

print('홍길동->김철수 계좌이체')
try:
    a1.transfer(600,a2)
except InvalidTransactionException as e:
    print(e)
print(a1.showAccountInfo())
print(a2.showAccountInfo())

bm = BankManager()
bm.addAccount(a1)
bm.addAccount(a2)
print('11-11 계좌 검색')
for i in range(len(bm.getAllAccountList())):
    print
r = bm.searchAccount('11-11')
if r == None:
    print('검색 결과가 없습니다.')
else:
    print(r.showAccountInfo())

소유주: 홍길동, 계좌번호: 11-11, 잔고: 500
소유주: 김철수, 계좌번호: 11-12, 잔고: 1000
홍길동->김철수 계좌이체
이체 잔고가 부족합니다.
소유주: 홍길동, 계좌번호: 11-11, 잔고: 500
소유주: 김철수, 계좌번호: 11-12, 잔고: 1000
11-11 계좌 검색
소유주: 홍길동, 계좌번호: 11-11, 잔고: 500


In [ ]:
from bank import Account, BankManager, InvalidTransactionException

def print_main_menu():
    print('---------------------')
    print('Welcome to DIMA Bank')
    print('---------------------')
    print('1. 계좌 생성')
    print('2. 입금')
    print('3. 출금')
    print('4. 이체')
    print('5. 전체 계좌 정보 출력')
    print('6. 종료')


manager = BankManager()
manager.loadData()
while True:
    print_main_menu()
    num = input('메뉴 선택:')
    if num == '1':
        name = input('이름: ')
        accNo = input('계좌 번호: ')
        balance = int(input('잔고: '))
        a = Account(name, accNo, balance)
        manager.addAccount(a)
        print('생성 완료되었습니다.')
    elif num == '2':
        accNo = input('입금할 계좌 번호: ')
        amount = int(input('입금 금액: '))
        a = manager.searchAccount(accNo)
        if a == None:
            print(f'{accNo} 계좌가 존재하지 않습니다.')
        else:
            a.deposit(amount)
            print('입금 완료되었습니다. ', a.showAccountInfo())
    elif num == '3':
        accNo = input('계좌 번호를 입력하세요: ')
        amount = int(input('출금 금액: '))
        a = manager.searchAccount(accNo)
        if a == None:
            print(f'{accNo} 계좌가 존재하지 않습니다.')
        else:
            try:
            a.withdraw(amount)
            print('출금 완료되었습니다. ', a.showAccountInfo())
            except InvalidTransactionException as e:
                print(e)
    elif num == '4':
        from_accNo = input('출금 계좌 입력: ')
        to_accNo = input('이체 계좌 입력: ')
        amount = int(input('이체 금액: '))
        from_a = manager.searchAccount(from_accNo)
        to_a = manager.searchAccount(to_accNo)
        if from_a != None and to_a != None:
            try:
                from_a.transfer(amount, to_a)
                print('이체 완료', from_a.showAccountInfo())
            except InvalidTransactionException as e:
                print(e)
        else:
            print('출금 또는 입금 계좌가 존재하지 않습니다.')
    elif num == '5':
        all_account = manager.getAllAccountList()
        for a in all_account:
            print(a.showAccountInfo())
    elif num == '6':
        manager.saveData()
        print('뱅킹 시스템 종료')
        break
    else: 
        print('메뉴 선택 오류')




# 자동 계좌 번호 생성 , 잔고 입력 예외 처리